### PR TITLE
Feature/plotting training progress

### PR DESCRIPTION
**Pull request by letiziaia**
- *timeGAN*: epochs and checkpoint frequency now defined in .fit(), all training losses are collected during training loops, optional synthetic data generation during .fit(), minor code clean-up
- *visualization*: added function to plot training losses, added option for feature averaging to tsne plots, code formatting (flake8 and black)

### CODE REVIEW

**Review by AlexanderVNikitin — CHANGES_REQUESTED** [review]

File: tsgm/models/timeGAN.py [hunk 1]
```diff
@@ -414,7 +422,23 @@ def _get_data_batch(self, data, n_windows: int):
             .repeat()
         )
 
-    def fit(self, data, epochs):
+    def fit(
+        self,
+        data: TensorLike,
+        epochs: int,
+        checkpoints_interval: int = None,
```
[the remark this answers is not in the document]
Comment (letiziaia) [reply tsgm/models/timeGAN.py]: this is done

File: tsgm/models/timeGAN.py [hunk 2]
```diff
@@ -414,7 +422,23 @@ def _get_data_batch(self, data, n_windows: int):
...
+        generate_synthetic: list = [],
```
[the remark this answers is not in the document]
Comment (letiziaia) [reply tsgm/models/timeGAN.py]: this is done

**Review by AlexanderVNikitin — CHANGES_REQUESTED** [review]

File: tsgm/models/timeGAN.py [hunk 3]
```diff
@@ -14,6 +14,42 @@
...
+        # If key is there, but value is not a list
+        except AttributeError:
+            # key -> [old_value, new_value]
+            super(LossTracker, self).__setitem__(key, [self[key], value])
+
+    def to_numpy(self) -> np.array:
+        """
+        :return 2d vector of losses
+        """
+        _losses = np.array([np.array(v) for v in self.values() if isinstance(v, list)])
```
[the remark this answers is not in the document]
Comment (letiziaia) [reply tsgm/models/timeGAN.py:44]: LossTracker now inherits from python ordered dictionary

File: tsgm/utils/visualization.py [hunk 4]
```diff
@@ -3,21 +3,29 @@
...
 
 
 DEFAULT_PALETTE_TSNE = {"hist": "red", "gen": "blue"}
 
 
-def visualize_dataset(dataset: tsgm.dataset.Dataset, obj_id: int = 0, palette=DEFAULT_PALETTE_TSNE, path: str = "/tmp/generated_data.pdf") -> None:
+def visualize_dataset(
+    dataset: tsgm.dataset.Dataset,
+    obj_id: int = 0,
+    palette=DEFAULT_PALETTE_TSNE,
```
[the remark this answers is not in the document]
Comment (letiziaia) [reply tsgm/utils/visualization.py]: done

File: tsgm/utils/visualization.py [hunk 5]
```diff
@@ -131,5 +204,56 @@ def visualize_original_and_reconst_ts(original, reconst, num=5, vmin=0, vmax=1):
...
     ids = np.random.choice(original.shape[0], size=num, replace=False)
     for i, sample_id in enumerate(ids):
-        axs[i, 0].imshow(original[sample_id].T, aspect='auto', vmin=vmin, vmax=vmax)
-        axs[i, 1].imshow(reconst[sample_id].T, aspect='auto', vmin=vmin, vmax=vmax)
+        axs[i, 0].imshow(original[sample_id].T, aspect="auto", vmin=vmin, vmax=vmax)
+        axs[i, 1].imshow(reconst[sample_id].T, aspect="auto", vmin=vmin, vmax=vmax)
+
+
+def visualize_training_loss(
+    loss_vector, labels=[], path: str = "/tmp/training_loss.pdf"
```
[the remark this answers is not in the document]
Comment (letiziaia) [reply tsgm/utils/visualization.py]: now it is empty tuple

**Pull request merged**